File: lib/generators/pusher.io/templates/pusher.io.js
var yaml = require('js-yaml'),
    fs = require('fs');

env = process.env.NODE_ENV || 'development'

var cfg = yaml.safeLoad(fs.readFileSync('./config/pusher.io.yml', 'utf8'))[env];

var crypto = require('crypto')
  , hmac
  , signature

var psignature = function(params) {
  hmac = crypto.createHmac("sha1", cfg.secret);
  hmac.update(params.join("\n"))
  return hmac.digest("hex");
}

var fs      = require('fs')
  , express = require('express')
  , app     = express()
  , server
  , io

if (cfg['ssl_key'] && cfg['ssl_cert']) {

  server = require('https').createServer({
    key: fs.readFileSync(cfg['ssl_key']),
    cert: fs.readFileSync(cfg['ssl_cert'])
  }, app);

} else {
  server = require('http').createServer(app);
}

io = require('socket.io')(server);

app.use(express.bodyParser());

app.post('/events', function(req, res) {
  b = req.body
  var params = [
    'POST',
    'events',
    b.app_id,
    b.channel,
    b.event,
    b.timestamp,
    cfg.key,
    cfg.host
  ];

  if (psignature(params) == b.signature) {
    io.sockets.in(b.channel).emit('pusher:' + b.event, b.data);
    res.send({});
  } else {
    res.send(401);
  }
});

app.post('/users', function(req, res) {
  if (req.body.key == cfg.key) {
    res.send({
      count: Object.keys(io.sockets.sockets).length
    });
  } else {
    res.send(401);
  }
});

io.use(function (socket, next) {
  var handshakeData = socket.handshake;
  var params = [
    'auth',
    handshakeData.query.app_id,
    handshakeData.query.timestamp,
    cfg.key
  ];
  if (psignature(params) == handshakeData.query.signature) {
    next();
  } else {
    next(new Error('not authorized'));
  }
});

io.on('connection', function (socket) {
  socket.on('subscribe', function(data) {
    socket.join(data.channel);
  });
  socket.on('unsubscribe', function(data) {
    socket.leave(data.channel);
  });
});

server.listen(cfg.port);

File: lib/pusher.io.rb
require 'pusher.io/version'
require 'openssl'
require 'net/http'
require 'net/https'

require "pusher.io/engine" if defined? Rails

module Pusher
  module IO

    class Channel
      attr_reader :name

      def initialize name
        @name = name.to_s
      end

      def trigger event, data={}
        cfg = Pusher::IO.config

        msg = {
          app_id: cfg[:app_id].to_s,
          channel: name,
          event: event.to_s,
          data: data,
          timestamp: Time.now.utc.to_i
        }

        sig = Pusher::IO.psignature([
          'POST',
          'events',
          msg[:app_id],
          msg[:channel],
          msg[:event],
          msg[:timestamp],
          cfg[:key],
          cfg[:host]
        ])

        msg[:signature] = sig

        form = Net::HTTP::Post.new('/events')
        form["Content-Type"] = "application/json"
        form.body = msg.to_json

        http = Net::HTTP.new(cfg[:host], cfg[:port])
        http.use_ssl = cfg[:ssl_key].present?
        http.start {|h| h.request(form)}
      end
    end

    class << self
      def config
        @config ||= begin
          cfg = {}
          filename = "#{Rails.root}/config/pusher.io.yml"
          yaml = YAML.load_file(filename)[Rails.env]
          raise ArgumentError, "The #{Rails.env} environment does not exist in #{filename}" if yaml.nil?
          yaml.each { |k, v| cfg[k.to_sym] = v }
          cfg
        end
      end

      def psignature(params)
        OpenSSL::HMAC.hexdigest('sha1', config[:secret], params.join("\n"))
      end

      def io_auth_query
        args = [
          'auth',
          config[:app_id],
          Time.now.utc.to_i,
          config[:key]
        ]
        sign = psignature(args)
        args << sign
        query = ""
        [:pusher, :app_id, :timestamp, :key, :signature].zip(args) do |key, value|
          query << "&#{key}=#{value}"
        end
        query
      end

      def socket_base_uri
        return config[:base_uri] if config[:base_uri]

        uri = config[:ssl_key].present? ? "https" : "http"
        uri << "://#{config[:host]}"
        uri << ":#{config[:port]}" if config[:port] && config[:port] != 80
        uri
      end

      def socket_uri
        socket_base_uri + "?#{io_auth_query}"
      end

      def socket_js_uri
        socket_base_uri + "/socket.io/socket.io.js"
      end

      def [] channel_name
        Channel.new channel_name
      end
    end
  end
end
